### PR TITLE
HIVE-28051: LLAP: cleanup local folders on startup and periodically

### PR DESCRIPTION
**Pull request by abstractdog**
### What changes were proposed in this pull request?
Implement a LocalDirCleaner which can remove old files from LLAP local dirs.

### Why are the changes needed?
Because LLAP cannot take care of cleaning up old files in every failure scenario, this was shown in a customer problem. When I investigated [HIVE-24272](https://issues.apache.org/jira/browse/HIVE-24272) I found that local files for failed queries/DAGs are cleaned up, so this is clearly about leftovers after daemon crashes.

### Does this PR introduce _any_ user-facing change?
No.

### Is the change a dependency upgrade?
No.

### How was this patch tested?
Unit test added, tested on llap daemon, added a file as:
```
mkdir -p  /apps/llap/work/usercache/hive/appcache/application_1707917402901_0001/3/output
touch  /apps/llap/work/usercache/hive/appcache/application_1707917402901_0001/3/output/file.out
```

set quick intervals like:
```
hive.llap.local.dir.cleaner.file.modify.time.threshold=60s
hive.llap.local.dir.cleaner.cleanup.interval=30s
```

checked logs:
```
query-executor <14>1 2024-02-15T07:37:59.351Z query-executor-0-0 query-executor 1 79184f83-fdb0-4982-9bd9-4e297438f8be [mdc@18060 class="impl.LocalDirCleaner" level="INFO" thread="pool-14-thread-1"] Cleaning up files older than 2024-02-15T07:36:59.351485Z from /apps/llap/work
query-executor <14>1 2024-02-15T07:38:29.351Z query-executor-0-0 query-executor 1 79184f83-fdb0-4982-9bd9-4e297438f8be [mdc@18060 class="impl.LocalDirCleaner" level="INFO" thread="pool-14-thread-1"] Cleaning up files older than 2024-02-15T07:37:29.351488Z from /apps/llap/work
query-executor <14>1 2024-02-15T07:38:59.351Z query-executor-0-0 query-executor 1 79184f83-fdb0-4982-9bd9-4e297438f8be [mdc@18060 class="impl.LocalDirCleaner" level="INFO" thread="pool-14-thread-1"] Cleaning up files older than 2024-02-15T07:37:59.351487Z from /apps/llap/work
query-executor <14>1 2024-02-15T07:38:59.352Z query-executor-0-0 query-executor 1 79184f83-fdb0-4982-9bd9-4e297438f8be [mdc@18060 class="impl.LocalDirCleaner" level="INFO" thread="pool-14-thread-1"] Delete old file: /apps/llap/work/usercache/hive/appcache/application_1707917402901_0001/3/output/file.out
query-executor <14>1 2024-02-15T07:39:29.351Z query-executor-0-0 query-executor 1 79184f83-fdb0-4982-9bd9-4e297438f8be [mdc@18060 class="impl.LocalDirCleaner" level="INFO" thread="pool-14-thread-1"] Cleaning up files older than 2024-02-15T07:38:29.351486Z from /apps/llap/work
query-executor <14>1 2024-02-15T07:39:59.351Z query-executor-0-0 query-executor 1 79184f83-fdb0-4982-9bd9-4e297438f8be [mdc@18060 class="impl.LocalDirCleaner" level="INFO" thread="pool-14-thread-1"] Cleaning up files older than 2024-02-15T07:38:59.351485Z from /apps/llap/work

```

### CODE REVIEW

**Comment by simhadri-g** [comment]
I think you missed updating the commit message with Hive jira

**Comment by abstractdog** [comment]
> I think you missed updating the commit message with Hive jira

LOL, thanks, fixed

**Comment by abstractdog** [comment]
unrelated test failure, I can rerun eventually, this can be reviewed

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
Comment (zhangbutao) [inline common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:5512]: Can we reuse the existing cleaning feature?
https://github.com/apache/hive/blob/e892ce7212b2f0c9e55092d89076b129c86c8172/common/src/java/org/apache/hadoop/hive/conf/HiveConf.java#L3894-L3898

Comment (abstractdog) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:5512]: I'll check: maybe the utility itself can be reused but with keeping a separate config

Comment (abstractdog) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:5512]: checked [ClearDanglingScratchDir](https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/session/ClearDanglingScratchDir.java), it seems to be specific to scratch dir cleanup and it's too smart for this scenario, I feel it would be more complicated to make it general and reuse than having a simple utility for the cases: ClearDanglingScratchDir checks lock files, works on hdfs, has a command line utility, etc., whereas LocalDirCleaner is just about to remove old files from a local folder

Comment (zhangbutao) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:5512]: Ok,will come back to check this after vacation. :)

File: llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java
Comment (zhangbutao) [inline llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java:111]: Maybe return a more user-friendly exception? And aslo this cleaner is not a very necessary thread, maybe we can just `LOG.warn/error` instead of throw exception?

File: llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java
Comment (zhangbutao) [inline llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java:48]: Is 1 thread enough to clean the old files?

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
Comment (zhangbutao) [inline common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:5517]: Nit: If the usual period is at least one day, can we use hour as the default unit?
`LLAP_LOCAL_DIR_CLEANER_FILE_MODIFY_TIME_THRESHOLD("hive.llap.local.dir.cleaner.file.modify.time.threshold", "24h", new TimeValidator(TimeUnit.HOURS),`

File: llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java
Comment (zhangbutao) [inline llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java:92]: I am not sure the how the loca dir is generated. So this maybe a question: Can we use `Files.list()` instread of `Files.walk()`?
Files.walk() will  retrieve all subdirectories recursively, and maybe a expencive operation if `pathLocalDir` has many sub-directories.

File: llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java
Comment (zhangbutao) [inline llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java:86]: Can be simplified by stream api:
```suggestion
localDirs.forEach(localDir -> cleanupPath(deleteBefore, Paths.get(localDir)));
```

File: llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java
Comment (abstractdog) [reply llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java:111]: right, logging is enough, we're not handling this exception anyway

File: llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java
Comment (abstractdog) [reply llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java:48]: yes, I think so, this is not something to be over-optimized, old files are not expected to be present at all, but if any, we don't have a requirement for the performance of this

File: llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java
Comment (abstractdog) [reply llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java:86]: good catch, I'll change

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
Comment (abstractdog) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:5517]: makes sense, I can change it to hour as default unit

File: llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java
Comment (abstractdog) [reply llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java:92]: we have a root folder like:
```
/apps/llap/work/
```

and we'll have dangling files like:
```
/apps/llap/work/usercache/hive/appcache/application_1707917402901_0001/3/output/attempt_1707917402901_0001_3_05_000002_0_10414:
total 16
drwxr-xr-x  2 hive hive   44 Feb 14 13:41 .
drwxr-xr-x 63 hive hive 4096 Feb 14 13:41 ..
-rw-r--r--  1 hive hive  425 Feb 14 13:41 file.out
-rw-r--r--  1 hive hive   32 Feb 14 13:41 file.out.index

/apps/llap/work/usercache/hive/appcache/application_1707917402901_0001/3/output/attempt_1707917402901_0001_3_05_000005_0_10416:
total 16
drwxr-xr-x  2 hive hive   44 Feb 14 13:41 .
drwxr-xr-x 63 hive hive 4096 Feb 14 13:41 ..
-rw-r--r--  1 hive hive  383 Feb 14 13:41 file.out
-rw-r--r--  1 hive hive   32 Feb 14 13:41 file.out.index

```

so the point is to find all dangling intermediate files which might be a few layers below then the root work dir, so Files.walk()mseemed to be the way to find them all and I'm a bit concerned about how complicated it would be to achieve the same with Files.list()

Comment (zhangbutao) [reply llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java:92]: Can `/apps/llap/work/usercache/hive/appcache/` be the root folder? Can we only check the first directory layer?
List the root folder  `/apps/llap/work/usercache/hive/appcache/` and get all the sub-directory whose name starts with `application_`?
new File("/apps/llap/work/usercache/hive/appcache/").listFiles(fn -> fn.getName().startsWith("application_"));

Comment (abstractdog) [reply llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java:92]: delete an application folder might be risky: for long-running apps it's possible that we're using the same application folder for a day, so a modification/creation date doesn't really help there, I cannot really think of a solution where we can safely an easily remove unused folders

Comment (zhangbutao) [reply llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java:92]: yes, might be risky. There seems to be no better way.

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
Comment (zhangbutao) [inline common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:5520]: nit:
```suggestion
      "Defaults to 24h, which is a reasonable period for a local file to be considered as a stale one."),
```

File: llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java
Comment (zhangbutao) [inline llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java:75]: nit:
```suggestion
        TimeUnit.HOURS);
```

**Review by zhangbutao — APPROVED** [review]
LGTM +1
Left several nit comments. Once the commets are fixed, we can merge this change without waitting CI.

File: llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java
Comment (zhangbutao) [inline llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java:48]: Maybe
`private long fileModifyTimeThresholdHour;`
or
`private long fileModifyTimeThreshold;`

File: llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java
Comment (zhangbutao) [inline llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java:57]: Maybe the LOG also should use `hour `unit.

File: llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java
Comment (zhangbutao) [inline llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java:83]: can use `HOUR `unit ?
If we want use `HOUR `as unit, should  we use `HOUR`  anywhere?

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java
Comment (abstractdog) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:5520]: right, I forgot to update this one :)

File: llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java
Comment (abstractdog) [reply llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java:75]: we cannot use this as HOUR default, because that way, every config with a shorter period will truncated to 0h:
```
2024-02-23T23:03:24,716  INFO [main] impl.LocalDirCleaner: Initialized local dir cleaner: interval: 0h, threshold: 0h
```
this happened when I tried to test this (in the unit test I need to configure SECONDS to get a result in a timely manner)
```
    HiveConf.setTimeVar(hiveConf, HiveConf.ConfVars.LLAP_LOCAL_DIR_CLEANER_CLEANUP_INTERVAL, 2, TimeUnit.SECONDS);
    HiveConf.setTimeVar(hiveConf, HiveConf.ConfVars.LLAP_LOCAL_DIR_CLEANER_FILE_MODIFY_TIME_THRESHOLD, 1,
        TimeUnit.SECONDS);
```

I think we can change the default unit in HiveConf to HOURS, but we're free to use SECONDS inside LocalDirCleaner, as it's the class' internal behavior

Comment (abstractdog) [reply llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java:75]: let's agree on this, and then we can resolve other comments I think

File: llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java
Comment (abstractdog) [reply llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java:48]: this depends on another review comment, I think we'll keep this SECONDS, so we can choose between fileModifyTimeThresholdSec and fileModifyTimeThreshold

File: llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java
Comment (abstractdog) [reply llap-server/src/java/org/apache/hadoop/hive/llap/daemon/impl/LocalDirCleaner.java:83]: HOUR is just the default unit for config values like "2"
the unit used here should correspond to the minus operation's first argument

**Comment by abstractdog** [comment]
@zhangbutao : thanks for you comments, according to the last comment, I'm considering this as approved, let me know if it's otherwise

**Comment by zhangbutao** [comment]
> @zhangbutao : thanks for your comments, according to the last comment, I'm considering this as approved, let me know if it's otherwise

Yes, free to merge this change. :)

**Pull request merged**
